Update scheduler framework document (#19606)
kubernetes/website: content/en/docs/concepts/scheduling/scheduling-framework.md

> {{< figure src="/images/docs/scheduling-framework-extensions.png" title="scheduling framework extension points" >}}

Reviewer: The image still says PostFilter

Author: Good catch. Image updated.

### QueueSort {#queue-sort}

These plugins are used to sort Pods in the scheduling queue. A queue sort plugin

> essentially will provide a `Less(Pod1, Pod2)` function. Only one queue sort

Reviewer 2: optional tidying:
```suggestion
essentially provides a `Less(Pod1, Pod2)` function. Only one QueueSort
```

filter plugin marks the node as infeasible, the remaining plugins will not be
called for that node. Nodes may be evaluated concurrently.

> ## PreScore

Reviewer: One `#` missing

Author: Good catch.

```go
func NormalizeScores(scores map[string]int) {
    highest := 0
    for _, score := range scores {
        highest = max(highest, score)
    }
    for node, score := range scores {
        scores[node] = score*NodeScoreMax/highest
    }
}
```

> If any Normalize Score plugin returns an error, the scheduling cycle is

Reviewer 2: Should this be:
```suggestion
If any NormalizeScore plugin returns an error, the scheduling cycle is
```
?

Reviewer 2: (and the same elsewhere)

This is the last step in a scheduling cycle. Once a Pod is in the reserved
state, it will either trigger [Unreserve](#unreserve) plugins (on failure) or
[PostBind](#post-bind) plugins (on success) at the end of the binding cycle.

{{< note >}}

> This concept used to be referred to as "assume".

Reviewer 2: When did it change / do readers need to know?

Author: "assume" is a term used internally. So let me remove this section to avoid confusion.

{{< /note >}}

### Permit

> Permit plugins are invoked at the end of the scheduling cycle for each pod to

Reviewer 2: nit:
```suggestion
_Permit_ plugins are invoked at the end of the scheduling cycle for each Pod, to
```

[PostBind](#post-bind) plugins (on success) at the end of the binding cycle.

### Permit

_Permit_ plugins are invoked at the end of the scheduling cycle for each Pod, to
prevent or delay the binding to the candidate node. A permit plugin can do one of
the three things:

1.  **approve** \
    Once all Permit plugins approve a Pod, it is sent for binding.

1.  **deny** \
    If any Permit plugin denies a Pod, it is returned to the scheduling queue.
    This will trigger [Unreserve](#unreserve) plugins.

1.  **wait** (with a timeout) \
    If a Permit plugin returns "wait", then the Pod is kept in an internal "waiting"
    Pods list, and the binding cycle of this Pod starts but directly blocks until it
    gets [approved](#frameworkhandle). If a timeout occurs, **wait** becomes **deny**
    and the Pod is returned to the scheduling queue, triggering [Unreserve](#unreserve)
    plugins.

{{< note >}}
While any plugin can access the list of "waiting" Pods and approve them
(see [`FrameworkHandle`](#frameworkhandle)), we expect only the permit
plugins to approve binding of reserved Pods that are in "waiting" state. Once a Pod

> is approved, it is sent to the [PreBind](#pre-bind) phase.

Reviewer 2: It's slightly better to [address the reader as “you”](https://kubernetes.io/docs/contribute/style/style-guide/#address-the-reader-as-you) _but_ let's not have that detail block a merge.

### Permit

> The Permit extension point contains two parts which spans scheduling cycle and

Reviewer: Are developers able to customize both parts?

Author: not for the second(internal) WaitOnPermit phase. I use a dotted rectangle to represent that in the drawing.

BTW: I've followed @ahg-g's comment to restructure this paragraph.

binding cycle.

In the scheduling cycle, Permit plugins are called to prevent or delay the binding

> of a Pod. A permit plugin can do one of three things:

Reviewer 3: ```suggestion
Permit plugins are invoked at the end of the scheduling cycle of each pod to prevent or delay the binding to the candidate node. A permit plugin can do one of three things:
```

    This will trigger [Unreserve](#unreserve) plugins.

1.  **wait** (with a timeout) \

>     If a permit plugin returns "wait", then the Pod is kept into an internal "waiting"

Reviewer 3: ```suggestion
    If a permit plugin returns "wait", then the Pod is kept in an internal "waiting"
```

    Pods list until a [plugin approves it](#frameworkhandle). If a timeout occurs,
    **wait** becomes **deny** and the Pod is returned to the scheduling queue, triggering

>     [Unreserve](#unreserve) plugins.

Reviewer 3: ```suggestion
    Pods list, and the binding goroutine of this pod starts but directly blocks until it gets [approved](#frameworkhandle). If a timeout occurs,
    **wait** becomes **deny** and the Pod is returned to the scheduling queue, triggering
    [Unreserve](#unreserve) plugins.
```

Author: let's say "binding cycle".

> **Note:** While any plugin can access the list of "waiting" Pods from the cache and

Reviewer 3: ```suggestion
**Note:** While any plugin can access the list of "waiting" Pods and
```

approve them (see [`FrameworkHandle`](#frameworkhandle)) we expect only the permit

> plugins to approve binding of reserved Pods that are in "waiting" state.

Reviewer 3: ```suggestion
plugins to approve binding of reserved Pods that are in "waiting" state. Once a Pod is approved, it is sent to the pre-bind phase.
```

In the beginning of binding cycle, if the current Pod is in "waiting" Pods list, scheduler
framework waits for it to be finished - either approved, denied, or timed out. Only when
the Pod is approved, it is sent to the subsequent [PreBind](#prebind) phase; otherwise the binding

> cycle is aborted.

Reviewer 3: I would remove this paragraph.

Author: Done.

Pod. If a bind plugin chooses to handle a Pod, **the remaining bind plugins are
skipped**.

> ### PostBind

Reviewer 2: To keep existing inbound links, I would (nit):
```suggestion
### PostBind {#post-bind}
```

```go
type Plugin interface {
    Name() string
}

type QueueSortPlugin interface {
    Plugin
    Less(*v1.pod, *v1.pod) bool
}

type PreFilterPlugin interface {
    Plugin
    PreFilter(context.Context, *framework.CycleState, *v1.pod) error
}

// ...
```

> ## Plugin Configuration

Reviewer 2: Optional tidying:
```suggestion
## Plugin configuration
```

Plugins can be enabled in the scheduler configuration. Also, default plugins can
be disabled in the configuration. Starting from 1.18, there are a number of [default 
plugins](/docs/reference/scheduling/profiles/#scheduling-plugins) enabled for the
scheduling framework.

In addition to default plugins, you can also implement your own scheduling
plugins and get them configured along with default plugins. You can visit

> [scheduler-plugins](https://github.com/kubernetes-sigs/scheduler-plugins) for more details.

Reviewer 2: nit: consider moving this inside a (new) What's Next section:
```markdown
(end of main text)
{{% /capture %}}
{{% capture whatsnext %}}
(further reading, etc)
{{% /capture %}}

Author: I'd put them under this section as they more fit here.

```go
type Plugin interface {
    Name() string
}

type QueueSortPlugin interface {
    Plugin
    Less(*v1.pod, *v1.pod) bool
}

type PreFilterPlugin interface {
    Plugin
    PreFilter(context.Context, *framework.CycleState, *v1.pod) error
}

// ...
```

## Plugin configuration

Plugins can be enabled in the scheduler configuration. Also, default plugins can
be disabled in the configuration. Starting from 1.18, there are a number of [default 
plugins](/docs/reference/scheduling/profiles/#scheduling-plugins) enabled for the

> scheduling framework.

Reviewer 2: How about:
```suggestion
You can enable or disable plugins in the scheduler configuration. If you are using
Kubernetes v1.18 or later, most scheduling
[plugins](/docs/reference/scheduling/profiles/#scheduling-plugins) are in use and
enabled by default.
```

(if short on time, the wording as it stands is OK to ship)

In addition to default plugins, you can also implement your own scheduling
plugins and get them configured along with default plugins. You can visit
[scheduler-plugins](https://github.com/kubernetes-sigs/scheduler-plugins) for more details.

> Starting from 1.18, you can configure a set of plugins as a scheduler profile and then

Reviewer 2: optional tweak:
```suggestion
If you are using Kubernetes v1.18 or later, you can configure a set of plugins as a
_scheduler profile_ and then
```

> # Plugin Configuration

Reviewer: This should have an extra `#`

Author: Done.

Plugins can be enabled in the scheduler configuration. Also, default plugins can
be disabled in the configuration. Starting from 1.18, there are a number of [default 
plugins](/docs/reference/scheduling/profiles/#scheduling-plugins) enabled for the scheduling framework.

> The scheduler configuration can include configuration for plugins as well. Such

Reviewer: All of this should go to the new doc.
Could you rebase on top of my PR?

Reviewer: But other than that, this section should have some notes about adding plugins to the out-of-tree registry.

Author: > All of this should go to the new doc.
> Could you rebase on top of my PR?

Nice, let me simply link to the profile doc. Do you want me to remove all the samples below, and then you can include them in `docs/reference/scheduling/profiles.md`?

> But other than that, this section should have some notes about adding plugins to the out-of-tree registry.

Reworded to mention that "In addition to default plugins, you can also implement your own scheduling plugins and get them configured along with default plugins."

Reviewer: I'll do a followup PR once the previous one gets merged.
